config/config.go

```diff
@@ -0,0 +1,173 @@
+package config
+
+import (
+	"context"
+	"crypto/rand"
+
+	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+
+	logging "github.com/ipfs/go-log"
+	circuit "github.com/libp2p/go-libp2p-circuit"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	host "github.com/libp2p/go-libp2p-host"
+	ifconnmgr "github.com/libp2p/go-libp2p-interface-connmgr"
+	pnet "github.com/libp2p/go-libp2p-interface-pnet"
+	metrics "github.com/libp2p/go-libp2p-metrics"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	tptu "github.com/libp2p/go-libp2p-transport-upgrader"
+	filter "github.com/libp2p/go-maddr-filter"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var log = logging.Logger("p2p-config")
+
+// AddrsFactory is a function that takes a set of multiaddrs we're listening on and
+// returns the set of multiaddrs we should advertise to the network.
+type AddrsFactory = bhost.AddrsFactory
+
+// NATManagerC is a NATManager constructor.
+type NATManagerC func(inet.Network) bhost.NATManager
+
+// Config describes a set of settings for a libp2p node
+type Config struct {
+	Transports         []TptC
+	Muxers             []MsMuxC
+	SecurityTransports []MsSecC
+	ListenAddrs        []ma.Multiaddr
+	PeerKey            crypto.PrivKey
+	Peerstore          pstore.Peerstore
+	Protector          pnet.Protector
+	Reporter           metrics.Reporter
+	Relay              bool
+	RelayOpts          []circuit.RelayOpt
+	Insecure           bool
+	ConnManager        ifconnmgr.ConnManager
+	AddrsFactory       bhost.AddrsFactory
+	Filters            *filter.Filters
+	NATManager         NATManagerC
+}
+
+// NewNode constructs a new libp2p Host based from the Config.
+func (cfg *Config) NewNode(ctx context.Context) (host.Host, error) {
+	// Check this early. Prevents us from even *starting* without verifying this.
+	if pnet.ForcePrivateNetwork && cfg.Protector == nil {
+		log.Error("tried to dial with no Private Network Protector but usage" +
+			" of Private Networks is forced by the enviroment")
+		// Note: This is *also* checked the upgrader itself so it'll be
+		// enforced even *if* you don't use the libp2p constructor.
+		return nil, pnet.ErrNotInPrivateNetwork
+	}
+
+	// If no key was given, generate a random 2048 bit RSA key
+	privKey := cfg.PeerKey
+	if cfg.PeerKey == nil {
+		priv, _, err := crypto.GenerateKeyPairWithReader(crypto.RSA, 2048, rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		privKey = priv
+	}
+
+	// Obtain Peer ID from public key
+	pid, err := peer.IDFromPublicKey(privKey.GetPublic())
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new blank peerstore if none was passed in
+	ps := cfg.Peerstore
+	if ps == nil {
+		ps = pstore.NewPeerstore()
+	}
+	if !cfg.Insecure {
```

Comment:
Why do we keep lugging around the insecure option?

Reply:
We use it for debugging. It's actually really useful. However, it's designed to be an all or nothing "disable all security" option.

Reply:
fair enough! I can see it being useful in analysing wire protocols without encryption.